### PR TITLE
Network bootstrapper check duplicate node names

### PR DESCRIPTION
**Pull request by stanly-johnson**
👮🏻👮🏻👮🏻 !!!! DESCRIBE YOUR CHANGES HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [x] Have you run the unit, integration and smoke tests as described here? https://docs.corda.net/head/testing.html
- [ ] If you added/changed public APIs, did you write/update the JavaDocs?
- [ ] If the changes are of interest to application developers, have you added them to the changelog, and potentially release notes?
- [x] If you are contributing for the first time, please read the agreement in CONTRIBUTING.md now and add to this Pull Request that you agree to it.

I certify that my contribution is in accordance with the terms of http://developercertificate.org/

Thanks for your code, it's appreciated! :)

### CODE REVIEW

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt [hunk 1]
```diff
@@ -92,6 +98,23 @@ class NetworkBootstrapper {
...
 
+    /*the function checks for duplicate myLegalName in the all the *_node.conf files
+    All the myLegalName values are added to a HashSet - this helps detect duplicate values.
+    If a duplicate name is found the process is aborted with an error message
+    */
+    private fun checkForDuplicateLegalNames(directory: Path, cordappJars: List<Path>) {
+      val legalNames = HashSet<String>()
+      val confFiles = directory.list { it.filter { it.toString().endsWith("_node.conf") }.toList() }
+      for (confFile in confFiles) {
+        val legalName = confFile.readLines { it.findFirst().get() }
```
Comment (shamsasari) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt]: You're assuming the legal name entry is the first line in the conf file - that's not guaranteed to be the case. Instead parse the conf file and extract out the "myLegalName" option. Have a look at `gatherNotaryInfos` in this class.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt [hunk 2]
```diff
@@ -92,6 +98,23 @@ class NetworkBootstrapper {
         }
     }
 
+    /*the function checks for duplicate myLegalName in the all the *_node.conf files
+    All the myLegalName values are added to a HashSet - this helps detect duplicate values.
+    If a duplicate name is found the process is aborted with an error message
+    */
+    private fun checkForDuplicateLegalNames(directory: Path, cordappJars: List<Path>) {
+      val legalNames = HashSet<String>()
+      val confFiles = directory.list { it.filter { it.toString().endsWith("_node.conf") }.toList() }
```
Comment (shamsasari) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt]: The bootstrapper can work in two modes, either on a list of _node.conf files, like here, or on an existing set of node directories with the `node.conf` file in each dir. We'd want to have duplication check for both scenerios. The easy way to do that is to simply move when this method is called to after `generateDirectoriesIfNeeded` is called and look up the `node.conf` file. Again have a look at `gatherNotaryInfos` which does the same thing.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt [hunk 3]
```diff
@@ -92,6 +98,23 @@ class NetworkBootstrapper {
...
+      for (confFile in confFiles) {
+        val legalName = confFile.readLines { it.findFirst().get() }
+        if(!legalNames.add(legalName)){
+          println("Duplicate Node Found - ensure every node has a unique legal name");
+          throw Error("Duplicate Node Found");
```
Comment (shamsasari) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt]: `IllegalArgumentException` would be more correct. `Error` is for more abnormal issues that typically are not expected to occur.

Also include the name of the duplicate to aid the user.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt [hunk 4]
```diff
@@ -293,13 +316,4 @@ class NetworkBootstrapper {
         override fun rpcClientKryoPool(context: SerializationContext) = throw UnsupportedOperationException()
         override fun rpcServerKryoPool(context: SerializationContext) = throw UnsupportedOperationException()
     }
-
-    private object AMQPParametersSerializationScheme : AbstractAMQPSerializationScheme(emptyList()) {
```
Comment (shamsasari) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt]: Hmm, I wonder why we created a custom one here? Thanks for that.

**Comment by stanly-johnson** [comment]
@shamsasari changes done in #3182

**Comment by shamsasari** [comment]
Thanks for that. It's better to stick to the same PR when making changes, but that's OK for now. I'll close this.

**Pull request closed without merging**